### PR TITLE
scala/codegen Fat Dar support

### PR DESCRIPTION
**Pull request by leo-da**
Add Scala Codegen fat dar support.
See #109 

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

### CODE REVIEW

**Comment by leo-da** [comment]
@stephencompall-DA do you think it is worth keeping `CodeGen.Mode`? `Legacy` mode is gone. There is only `Novel` one. I believe it will be a bit cleaner without it.

**Comment by leo-da** [comment]
> @stephencompall-DA do you think it is worth keeping `CodeGen.Mode`? `Legacy` mode is gone. There is only `Novel` one. I believe it will be a bit cleaner without it.

**Comment by leo-da** [comment]
@robin-da @martin-drhu-da once this PR is merged you can remove the following targets:
- `"//daml-foundations/daml-tools/da-hs-damlc-app:da-hs-damlc-app-deprecated"`
- `//daml-foundations/daml-ghc/package-database/deprecated:daml-prim-1.3.dalf"`
assuming no other builds are using them.

File: language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/CodeGen.scala [hunk 1]
```diff
@@ -109,78 +55,81 @@ object CodeGen {
...
-      mode: Mode): String \/ Seq[mode.InterfaceElement] =
-    urls
-      .map { url =>
-        val is = url.openStream()
-        try decodePackageFrom(is, mode)
-        finally is.close()
-      }
-      .toList
-      .sequenceU
+    val errorsOrUnits: ValidationNel[String, List[Unit]] = files.traverseU { f =>
```
Comment (stephencompall-DA) [inline language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/CodeGen.scala]: I believe this traversal will create separate runs for each sub-`EnvironmentInterface`, producing multiple EventDecoder and package ID files for multi-argument situations, probably overwriting as it goes. If so, we need to handle the whole argument set with a single `decodeInterface` to produce a single `EnvironmentInterface` to do the whole codegen run on.

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/Dar.scala [hunk 2]
```diff
@@ -14,4 +16,14 @@ object Dar {
     override def map[A, B](fa: Dar[A])(f: A => B): Dar[B] =
       Dar[B](main = f(fa.main), dependencies = fa.dependencies.map(f))
   }
+
+  implicit val darTraverse: Traverse[Dar] = new Traverse[Dar] {
```
[the remark this answers is not in the document]
Comment (leo-da) [reply daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/Dar.scala]: yeah, I thought about actually removing the functor completely... `map` can be derived from `traverse`.

I actually like passing the typeclasses explicitly. More than 1 typeclass instance is possible, right? If we used newtypes for every instance... like you suggested a while ago... I would not mind giving it some obscure name

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/Dar.scala [hunk 3]
```diff
@@ -14,4 +16,14 @@ object Dar {
...
+    override def traverseImpl[G[_]: Applicative, A, B](fa: Dar[A])(f: A => G[B]): G[Dar[B]] = {
+      val G: Applicative[G] = implicitly
+      val gb: G[B] = f(fa.main)
+      val L: Traverse[List] = scalaz.std.list.listInstance
+      val gbs: G[List[B]] = L.traverse(fa.dependencies)(f)
+      G.apply2(gb, gbs)((b, bs) => Dar(b, bs))
```
[the remark this answers is not in the document]
Comment (leo-da) [reply daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/Dar.scala]: yeah, I like this. But I would still use `G.apply2` more readable :)

File: language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/CodeGen.scala (hunk 1 above)
Comment (leo-da) [reply language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/CodeGen.scala]: this is something worth of a test case... I will create a second daml file for the Integration Test.

**Comment by garyverhaegen-da** [comment]
Should this add a line to the changelog?

**Comment by leo-da** [comment]
> Should this add a line to the changelog?

It is not part of the public SDK. So, I guess it should not be in the changelog to avoid confusion. We are planning to release it in the SDK at some point.

**Pull request merged**